### PR TITLE
Programs Grammar

### PR DESCRIPTION
**Pull request by YuriHechter**
https://docs.google.com/spreadsheets/d/1q7nKXTVKjg1cSPEU06Bov2WORnF9tCVt1u-9MEVZTWw/edit#gid=1849713197

### CODE REVIEW

File: cache/search_stats.go [hunk 1]
```diff
@@ -147,15 +147,18 @@ type SearchStatsCache interface {
 	// Some of the sources (consts.NOT_TO_INCLUDE_IN_SOURCE_BY_POSITION) are restricted from these functions so you should not use them for general porpuses.
 	GetSourceByPositionAndParent(parent string, position string, sourceTypeIds []int64) *string
 	GetSourceParentAndPosition(source string, getSourceTypeIds bool) (*string, *string, []int64, error)
+
+	GetProgramByCollectionAndPosition(collection_uid string, position string) *string
```
Comment (bbfsdev) [inline cache/search_stats.go:151]: Why position is string?

File: cache/search_stats.go [hunk 2]
```diff
@@ -199,10 +202,10 @@ func (ssc *SearchStatsCacheImpl) IsSourceWithEnoughUnits(uid string, count int,
 }
 
 func (ssc *SearchStatsCacheImpl) GetSourceByPositionAndParent(parent string, position string, sourceTypeIds []int64) *string {
-	if sourceTypeIds == nil || len(sourceTypeIds) == 0 {
+	if len(sourceTypeIds) == 0 {
 		sourceTypeIds = consts.ALL_SRC_TYPES
 	}
-	for typeId := range sourceTypeIds {
+	for _, typeId := range sourceTypeIds {
```
Comment (bbfsdev) [inline cache/search_stats.go:208]: That was a bug? we were using index instead of type id?

File: cache/search_stats.go [hunk 3]
```diff
@@ -472,3 +487,47 @@ func (ssc *SearchStatsCacheImpl) loadSourcesByPositionAndParent() (map[string]st
...
+		}
+		if collection_uid == consts.PROGRAM_COLLECTION_NEW_LIFE {
+			if val, err := strconv.Atoi(name); err == nil {
+				position = val
+			} else {
+				fmt.Printf("The value '%v' of column 'name' from 'collections_content_units' was expected to be numeric and present the chapter number for New Life program.\n", name)
+				continue
+			}
+		} else if collection_uid == consts.PROGRAM_COLLECTION_HAPITARON {
+			position--
```
Comment (bbfsdev) [inline cache/search_stats.go:527]: Why --?

File: cache/search_stats.go [hunk 4]
```diff
@@ -147,15 +147,18 @@ type SearchStatsCache interface {
...
-	conventions                map[string]map[string]int
-	holidayYears               map[string]map[string]int
-	sourcesByPositionAndParent map[string]string
+	mdb                             *sql.DB
+	tags                            ClassByTypeStats
+	sources                         ClassByTypeStats
+	conventions                     map[string]map[string]int
+	holidayYears                    map[string]map[string]int
+	sourcesByPositionAndParent      map[string]string
+	programsByCollectionAndPosition map[string]string
```
Comment (bbfsdev) [inline cache/search_stats.go:161]: What is this being used for?

Comment (bbfsdev) [inline cache/search_stats.go:161]: What is this being used for?

File: cache/search_stats.go [hunk 5]
```diff
@@ -472,3 +487,47 @@ func (ssc *SearchStatsCacheImpl) loadSourcesByPositionAndParent() (map[string]st
...
+	for rows.Next() {
+		var collection_uid string
+		var position int
+		var name string
+		var program_uid string
+		err = rows.Scan(&collection_uid, &position, &name, &program_uid)
+		if err != nil {
+			return nil, errors.Wrap(err, "rows.Scan")
+		}
+		if collection_uid == consts.PROGRAM_COLLECTION_NEW_LIFE {
```
Comment (bbfsdev) [inline cache/search_stats.go:519]: Why the special cases?

File: consts/consts.go [hunk 6]
```diff
@@ -525,15 +528,20 @@ var INTENT_HIT_TYPE_BY_CT = map[string]string{
 const (
 	GRAMMAR_INDEX = "grammar"
 
-	GRAMMAR_TYPE_FILTER         = "filter"
-	GRAMMAR_TYPE_LANDING_PAGE   = "landing-page"
-	GRAMMAR_TYPE_CLASSIFICATION = "classification"
+	GRAMMAR_TYPE_FILTER              = "filter"
+	GRAMMAR_TYPE_FILTER_WITHOUT_TERM = "filter_without_term"
```
Comment (bbfsdev) [inline consts/consts.go:532]: what filter without term means? Why is it here in this change?

File: consts/consts.go [hunk 7]
```diff
@@ -525,15 +528,20 @@ var INTENT_HIT_TYPE_BY_CT = map[string]string{
...
+	GRAMMAR_TYPE_LANDING_PAGE        = "landing-page"
+	GRAMMAR_TYPE_CLASSIFICATION      = "classification"
 
-	GRAMMAR_INTENT_FILTER_BY_CONTENT_TYPE       = "by_content_type"
-	GRAMMAR_INTENT_FILTER_BY_SOURCE             = "by_source"
-	GRAMMAR_INTENT_SOURCE_POSITION_WITHOUT_TERM = "source_position_without_term"
+	GRAMMAR_INTENT_FILTER_BY_CONTENT_TYPE         = "by_content_type"
+	GRAMMAR_INTENT_FILTER_BY_SOURCE               = "by_source"
+	GRAMMAR_INTENT_FILTER_BY_PROGRAM              = "by_program"
+	GRAMMAR_INTENT_FILTER_BY_PROGRAM_WITHOUT_TERM = "by_program_without_term"
```
Comment (bbfsdev) [inline consts/consts.go:539]: Not sure why all those filters needed/used now.

File: consts/consts.go [hunk 8]
```diff
@@ -666,7 +689,9 @@ const (
 	VAR_CONTENT_TYPE        = "$ContentType"
 	VAR_SOURCE              = "$Source"
 	VAR_POSITION            = "$Position"
-	VAR_DIVISION_TYPE       = "$DivType"
+	VAR_DIVISION_TYPE       = "$DivisionType"
+	VAR_PROGRAM             = "$Program"
+	VAR_RESTRICTED          = "$Restricted" // Search terms that privent triggering grammar engine.
```
Comment (bbfsdev) [inline consts/consts.go:694]: What are those? How are they used?

File: search/variables/content_type.variable [hunk 9]
```diff
@@ -152,6 +153,7 @@ ru,lessons => урок рава
 ru,lessons => уроки рава
 ru,lessons => урок с равом
 ru,lessons => уроки с равом
+ru,lessons => УУ
```
Comment (bbfsdev) [inline search/variables/content_type.variable:156]: Nice

File: search/variables/program.variable [hunk 10]
```diff
@@ -0,0 +1,11 @@
+he,t9oGIcN5 => עולם
```
Comment (bbfsdev) [inline search/variables/program.variable:1]: Nice, are those in addition to what we have in the database?

File: search/variables/restricted.variable [hunk 11]
```diff
@@ -0,0 +1,17 @@
+en,psalms => psalms
```
Comment (bbfsdev) [inline search/variables/restricted.variable:1]: Not sure how that is being used.

File: search/grammar_variables_matcher.go [hunk 12]
```diff
@@ -16,141 +18,295 @@ func GrammarFilterVariablesMatch(intent string, variablesByPhrase VariablesByPhr
...
-					return false
-				}
-				hasVarSource = true
-			}
-		}
-		if !(hasVarText && hasVarSource) {
-			log.Warningf("Filter intent by content type must have one appearance of $Text and one appearance of $Source")
-			return false
-		}
+	switch intent {
```
Comment (bbfsdev) [inline search/grammar_variables_matcher.go:21]: You just refactored the code here? Any new code?

File: search/grammar_query.go [hunk 13]
```diff
@@ -240,3 +249,16 @@ func chooseRule(query *Query, rules []string) string {
 	}
 	return rules[maxIndex]
 }
+
+func hasCommonFilter(a map[string][]string, b map[string][]string) bool {
+	for filterName, values := range a {
+		sort.Strings(values)
```
Comment (bbfsdev) [inline search/grammar_query.go:255]: Sort values only if filterName is in b.

File: search/grammar_v2_engine.go [hunk 14]
```diff
@@ -279,36 +286,48 @@ func (e *ESEngine) SearchByFilterIntents(filterIntents []Intent, filters map[str
...
-					resultsByLang[intent.Language] = FilteredSearchResult{
-						Results:     results,
-						Term:        text,
-						ContentType: contentType,
-						HitIdsMap:   hitIdsMap,
-						MaxScore:    maxScore,
+					resultByLang := FilteredSearchResult{
+						Results:                  results,
+						Term:                     text,
+						PreserveTermForHighlight: programCollection != "",
```
Comment (bbfsdev) [inline search/grammar_v2_engine.go:323]: Why we cannot have term for highlight in programCollection?

File: search/grammar_v2_engine.go [hunk 15]
```diff
@@ -387,11 +406,17 @@ func (e *ESEngine) searchResultsToIntents(query *Query, language string, result
...
 	minScoreByLandingPage := make(map[string]float64)
-	addPositionWithoutTerm := true
-	for filterKey := range query.Filters {
-		if _, ok := consts.AUTO_INTENTS_BY_SOURCE_NAME_SUPPORTED_FILTERS[filterKey]; !ok {
-			addPositionWithoutTerm = false
-			break
+	queryTermIsNumber, queryTermHasDigit := utils.HasNumeric(query.Term)
+	// In case our query is numeric only, we ignore intents of "program\source position without term" to avoid irrelavnt results.
+	// Also we support "program with position without term" intents only if we have a numeric chapter as part of the query.
+	addProgramPositionWithoutTerm := !queryTermIsNumber && queryTermHasDigit
```
Comment (bbfsdev) [inline search/grammar_v2_engine.go:412]: Don't understand, lets talk about this.

File: search/grammar_v2_engine.go [hunk 16]
```diff
@@ -948,3 +1065,105 @@ func (e *ESEngine) sourceUidByTerm(term string, languages []string) (*string, st
...
+				if termLC == strings.ToLower(resTerm) {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func (e *ESEngine) selectFilterIntent(intents []Intent) (*Intent, error) {
```
Comment (bbfsdev) [inline search/grammar_v2_engine.go:1084]: Not sure what is that needed for.

File: search/grammar_v2_engine.go [hunk 17]
```diff
@@ -948,3 +1065,105 @@ func (e *ESEngine) sourceUidByTerm(term string, languages []string) (*string, st
 	}
 	return nil, "", false
 }
+
+func (e *ESEngine) isTermRestricted(term string, languages []string) bool {
```
Comment (bbfsdev) [inline search/grammar_v2_engine.go:1069]: Not like this as you don't tokenize anything here...
We should try tokenize everything in elastic.

**Review by bbfsdev — CHANGES_REQUESTED** [review]

File: search/grammar_v2_engine.go (hunk 17 above)
Comment (YuriHechter) [reply search/grammar_v2_engine.go:1069]: Add comment

**Pull request merged**